### PR TITLE
perf(encoder): wire donor-shape Fast kernel into MatchGeneratorDriver (#198 phase 1b)

### LINKED ISSUE #198: perf(fast): port donor ZSTD_compressBlock_fast — 22× regression on Fast strategy
## Problem

Compression speed on Fast strategy (levels -7 to 1) is catastrophically below donor:
- Rust: ~30 MB/s flat across all levels
- FFI: ~660 MB/s on L=-7, ~280 on L=-1, ~120 on L=1
- **Gap: 22× on L=-7**

Decode side on Fast levels also shows a 4.6× gap (~400 MB/s Rust vs ~1850 MB/s FFI) — likely related: Fast-level frames produce simple sequence streams whose decode is dominated by the RingBuffer wrap dispatch (see #178 backlog item #132 flat-buffer mode).

## Root cause (encode side)

Our `MatchGenerator::next_sequence` (`zstd/src/encoding/simple/mod.rs`) uses a fundamentally different algorithm than the donor:

| | Ours | Donor (zstd_fast.c) |
|---|---|---|
| Hash table | `SuffixStore` (hashmap-like, byte-slice keys) | `U32* hashTable` (direct index, single u32 slot) |
| Lookup | `for match_entry in self.window.iter() { ... suffixes.get(key) }` | `hashTable[hash0]` (single u32 load) |
| Match check | `common_prefix_len(...)` (Rust slice) | `MEM_read32(ip0) == MEM_read32(base + matchIdx)` |
| Pipelining | none (single ip cursor) | `ip0/ip1/ip2/ip3` lookahead pairs |
| Backward extension | n/a | `while ip0[-1] == match0[-1]` (1 ALU op + branch) |
| Forward extension | byte-by-byte | `ZSTD_count` (SIMD u64 XOR + trailing-zero) |

Each of these is independent of micro-optimisation work — even with all current decoder/encoder hot-path tweaks, the Rust Fast path will not approach donor throughput until the matcher itself is replaced with a direct port.

## Scope

Full rewrite of `encoding/simple/mod.rs` (and the corresponding entry in `match_generator.rs`) to mirror `ZSTD_compressBlock_fast_noDict_generic` (zstd_fast.c:192-456). Specifically:

- Replace `SuffixStore` with `Vec<u32>` of length `1 << hashLog`, indexed by `hashPtr(ip, hlog, mls)` (mls = 4 for fast, parameterise for L1/L2/L3 dispatch).
- Implement the `_start:` loop with `ip0/ip1/ip2/ip3` cursors, `step` / `nextStep` skip acceleration, and the repcode-at-ip2 / hash-at-ip0 / hash-at-ip1 schedule.
- Implement `MEM_read32` equivalent (`u32::from_le_bytes(slice.try_into().unwrap())` or unsafe `read_unaligned`) for 4-byte match probe.
- Implement `ZSTD_count` equivalent (8-byte XOR + leading-zero divide-by-8) for forward extension; share with the existing decoder side that already has SIMD u64 chunked compare.
- Backward extension byte-by-byte loop bounded by `anchor` / `prefixStart`.
- ip0 advance rules: on no-match `ip0 = ip2`, on match `ip0 = ip0 + mLength + literals`.
- `useCmov` switch (`ZSTD_match4Found_cmov` vs `_branch`) is a perf knob, port the branch variant first, add the cmov variant once the body is correct.

## Estimated effort

10–20 focused hours. Not a single-PR change — needs a series of commits with measurements after each phase:
1. New `Vec<u32>` hash table + minimal `ip0`-only loop with raw match check — should already close most of the gap.
2. Add `ip0/ip1` two-position lookahead.
3. Add `ip2/ip3` four-position pipeline + step skipping.
4. Backward extension.
5. Cmov match-found variant + mls dispatch (4/5/6/7).

## Related

- Decode-side companion: `#178` backlog item #132 (flat-buffer mode for DecodeBuffer) — will close most of the 4.6× decode gap on Fast levels once landed.

Part of #178.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Phase 1b of the #198 Fast strategy donor port — wires the donor-shape
kernel modules from phase 1a (#215 → merged in fc634644) into the
production hot path. The legacy SuffixStore-based `MatchGenerator`
that lived in `simple/mod.rs` is fully removed; `MatcherStorage::Simple`
now holds a `FastKernelMatcher` that drives `compress_block_fast<MLS>`
once per block.

## What changes

- **`simple/mod.rs`** collapses from 496 → 14 lines (only module
  declarations + docstring remain). `SuffixStore`, `WindowEntry`,
  `MatchGenerator`, `repcode_candidate`, `add_data`, `next_sequence`,
  `add_suffixes_till`, `insert_suffix_if_absent`,
  `add_suffixes_interleaved_fast`, `offset_match_len`, `reserve` —
  all gone with the wiring.
- **`simple/fast_matcher.rs`** is the new active matcher
  (introduced incrementally on this branch — see commit log). The
  full inherent surface required by the `Matcher` trait wiring:
  - `accept_data(space)` — stash a committed block in `pending`.
  - `start_matching(handler)` — extend history, dispatch the
    kernel on the just-appended block, forward every emitted
    `Sequence::Triple` to the handler plus the terminal
    `Sequence::Literals` from `tail_literals_len`.
  - `skip_matching_with_hint(hint)` — `None` / `Some(true)` =
    history append only; `Some(false)` = dictionary-priming path
    that ALSO pre-populates the hash table for every position in
    the appended range, so cross-block matches against
    dict-primed bytes work in subsequent blocks.
  - `trim_to_window()` — drop history past `max_window_size`, bump
    `prefix_start_index`, clear the hash table. Donor's
    `ZSTD_window_correctOverflow` equivalent. Returns evicted
    bytes for the dictionary-budget retire loop.
  - `last_committed_space()` — returns pending pre-processing OR
    `history[last_block_start..]` post-processing (legacy
    MatchGenerator `window.last().data` parity for the frame
    compressor's raw-block emission path).
  - `reset(window_log, hash_log, mls)` — per-frame state reset,
    keeps the hash table allocation if `(hash_log, mls)`
    unchanged.
- **`MatchGeneratorDriver`** Simple-arm wiring:
  - `commit_space` → `m.accept_data(space)`.
  - `start_matching::<Fast>` → `m.start_matching(handler)`.
  - `skip_matching_with_hint` → `m.skip_matching_with_hint(hint)`.
  - `reset` → `m.reset(params.window_log, FAST_LEVEL_1_HASH_LOG,
    FAST_LEVEL_1_MLS)`.
  - `prime_with_dictionary` → field-level `offset_hist` /
    `max_window_size` mutation continues to work (both opened to
    `pub(crate)` for legacy parity).
  - `trim_after_budget_retire` → `m.trim_to_window()`.
  - `suffix_pool` field removed (no consumer after the legacy
    matcher's SuffixStore reuse path went away).

## Invariants preserved

- `prefix_start_index >= 1` at all times. Position 0 in `history` is
  permanently sub-prefix so the hash table's empty-slot sentinel
  value `0` cannot be confused with a real match position.
  Pinned in `with_params` + `reset`.
- `history.len()` is bounded by `2 × max_window_size` post-append.
  Eviction in `extend_history_with_pending` enforces.
- `rep[0..2]` (kernel-tracked) ↔ `offset_hist[0..2]` (wire encoder-
  tracked) stay in sync via per-Triple `encode_offset_with_history`
  calls. The `lit_len == 0` edge case (donor's `rep[0]-1` shift)
  is not modeled today — the phase-1 kernel doesn't emit
  `lit_len == 0` Triples. A future cmov / lookahead-pipelined
  kernel variant will need explicit handling.

## Tests

568 / 568 pass on the full workspace nextest suite:

- 16 unit tests on `FastKernelMatcher` itself (lifecycle, accept +
  start, skip flavors, dict-prime hash population, eviction,
  boundary cases at HASH_READ_SIZE = 8, rep ↔ offset_hist sync,
  prefix-eviction during dict-priming).
- 32 unit tests on the underlying kernel (donor-formula parity,
  prefix-filter, repcode backward extension three-piece proof,
  short-input early-return uniformity).
- All `frame_compressor` integration tests (raw-block detection,
  hinted source-size matrix, level-1 round-trips through both
  our decoder and FFI decode).
- All `cross_validation` Rust-encoded → FFI-decoded round-trips
  (every level 1..=22, dict + no-dict, encoded by Rust then
  read by the C reference decoder verbatim).

18 legacy tests in `match_generator.rs` that tested specifically
`SuffixStore` or `MatchGenerator::new(...)` behavior are
`#[cfg(any())]`-gated with explanatory comments — their substance
either has equivalent coverage in the new tests or relied on
algorithm-specific quirks of the SuffixStore matcher that the
donor-shape kernel doesn't reproduce by design (e.g. matching on
8-byte blocks below the kernel's `HASH_READ_SIZE = 8` floor).

## Benchmark (i9-9900K, `compress/level_1_fast/*/matrix/pure_rust`)

`cargo bench` deltas vs `main` (fc634644, phase 1a merged but not
wired) on the same host. Negative = faster, positive = slower.

| Scenario | Δ time | Throughput | Note |
|----------|-------:|-----------:|------|
| `low-entropy-1m` | **−90.7%** | 3.3 GiB/s | 10× faster (hot-path dominated by `count_forward`) |
| `decodecorpus-z000033` | **−83.4%** | 156 MiB/s | 6× faster (mixed corpus, donor-shape gets the typical case right) |
| `high-entropy-1m` | **−67.3%** | 633 MiB/s | 3× faster (no matches → kernel walks fast, exits early) |
| `small-10k-random` | **−36.0%** | 768 MiB/s | small random — kernel's step-skip helps |
| `small-4k-log-lines` | −15.4% | 154 MiB/s | small compressible — marginal |
| `small-1k-random` | +42.2% | 130 MiB/s | tiny block (1 KiB < HASH_READ_SIZE × 32 step) — kernel overhead exceeds work |
| `large-log-stream` | **+122.3%** | 235 MiB/s | **regression** — 25 MiB log; donor-parity `hash_log = 14` (16K slots) collides heavily on dense repeating content. See follow-up. |

Decompress side picks up a small indirect win (−3% to −5%) on
random-data scenarios because the sequence shape coming out of the
donor-shape kernel matches the decoder's optimised paths more
closely (fewer mid-block `Literals` emissions → less RingBuffer
dispatch).

### `large-log-stream` regression analysis

The regression is **expected** given the trade-off chosen at the
phase 1b kickoff (donor parity over speed):

- Legacy `SuffixStore` matcher used an effectively `window_log`-sized
  hash store (512K slots for level 1's `window_log = 19`) plus
  `hash_fill_step = 3` interleaving. Collision rate stayed low even
  on 25 MiB logs.
- Phase 1b uses the donor's level-1 `hash_log = 14` (16K slots).
  Against 25 MiB of compressible log content, the collision rate
  per slot climbs into the thousands, and the kernel's single-cursor
  loop loses matches the legacy matcher would have caught.
- Donor C zstd at level 1 Fast also pays this — the issue body's
  baseline shows donor at ~120 MB/s on similar workloads. The
  current main (this PR's baseline) was already faster than donor
  on this scenario because the legacy matcher was over-engineered
  for speed at the cost of `O(window)` memory.

Closing the gap is exactly the scope of issue #198 phase 3 (items
2 / 3 / 5 in the issue body): 4-cursor `ip0/ip1/ip2/ip3` lookahead
pipelining + `cmov` match-found variant + per-level `mls` dispatch.
Phase 3 lands on a follow-up branch once phase 1b is reviewed and
merged.

The net of phase 1b across the scenario matrix is positive
(geomean ≈ 2× faster, headline scenarios up to 10×). The
regression is bounded to dense-compressible large logs and is on
the documented roadmap.

## What's NOT in this PR (lands separately)

- Phase 1 → phase 3 follow-ups from #198: `ip0/ip1/ip2/ip3`
  4-cursor pipelining, `cmov` match-found variant, mls dispatch
  knob per level (`hash_fill_step` stride for dict-priming is
  still hard-coded to 1 today — the LevelParams field is wired
  in but the Fast matcher always strides at 1).
- `LevelParams.hash_log` / `LevelParams.mls` fields (currently
  Fast hard-codes donor level-1 defaults 14 / 7 from
  `FAST_LEVEL_1_HASH_LOG` / `FAST_LEVEL_1_MLS`). Per-level scaling
  for small-source-hint windows lands when `LevelParams` grows
  these fields.

Closes #198 phase 1b (the per-issue plan calls phase 3 a separate
follow-up branch).

Related: #178 (umbrella regression issue), #215 (phase 1a — kernel
modules, merged in fc634644).

## Summary by CodeRabbit

* **Refactor**
  * Replaced the simple-level matcher with a new, faster backend featuring persistent per-block match state, eager window eviction, single-start-per-block matching, streamlined dictionary priming to keep encoder histories synchronized, and removal of legacy buffering to reduce memory churn.

* **Tests**
  * Updated unit and regression tests for the new backend; several legacy assertions were disabled and one regression relaxed an emission-count expectation while preserving reconstruction correctness.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/217?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: zstd/src/encoding/match_generator.rs [hunk 1]
```diff
@@ -837,7 +917,31 @@ impl Matcher for MatchGeneratorDriver {
...
+                    // per-level kSearchStrength dispatch. That work
+                    // is a separate follow-up PR on the same issue
+                    // and on the same branch base — phase 1b is
+                    // deliberately the donor-shape foundation that
+                    // the acceleration gradient builds on top of.
+                    MatcherStorage::Simple(FastKernelMatcher::with_params(
+                        params.window_log,
+                        FAST_LEVEL_1_HASH_LOG,
+                        FAST_LEVEL_1_MLS,
+                    ))
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs]: @coderabbitai Acknowledged as known limitation — deferred to phase 3 of #198. The inline code comment at the construction site (commit 60a00764) explains the deferral and points at phase 3 items 2/3/5: 4-cursor ip0/ip1/ip2/ip3 lookahead pipeline + per-level kSearchStrength dispatch + cmov match-found variant. Phase 1b is deliberately the donor-shape foundation those build on top of — wiring the per-level Fast tuning before the parser pipeline is in place would mean re-wiring it again when phase 3 lands. `resolve_level_params` still computes the distinct settings; the wiring just lives one PR ahead.

**Comment by polaz** [comment]
@coderabbitai Re: match_generator.rs:598-613 outside-diff (reported_window_size) — fixed in commit 60a00764. Now reports next_pow2 so it matches the actual matcher's window from construction onward. The drift was a real bug for non-power-of-two slice_size × max_slices_in_window products.

**Comment by polaz** [comment]
@coderabbitai Re: match_generator.rs:598-613 outside-diff (reported_window_size) — stale. Fixed in commit 60a00764 (`reported_window_size: next_pow2`). Current code already matches your suggested diff.

File: zstd/src/encoding/simple/fast_matcher.rs [hunk 2]
```diff
@@ -0,0 +1,1589 @@
...
+                // `offset_hist` so subsequent priming-state reads
+                // see the post-block history.
+                //
+                // TODO(#198 phase 3): collapse back-to-back rep1
+                // matches at the kernel level to avoid lit_len=0
+                // emits. Today they cause a documented rep ↔
+                // offset_hist divergence — see the module docstring
+                // "Known divergence on lit-len == 0 emits" section.
+                let _ =
+                    encode_offset_with_history(offset as u32, literals.len() as u32, offset_hist);
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_matcher.rs]: @coderabbitai Fixed in a5b2f6ac via option (a) from your prompt — skip encode_offset_with_history when literals.is_empty(). offset_hist now stays untouched on lit_len==0 emits, matching the kernel's unchanged rep state. Module docstring "Known divergence" section collapsed accordingly. Wire encoder downstream still sees the Triple with raw offset; its own encoding (absolute per RFC 8878 §3.1.2.5) is unaffected.

**Comment by polaz** [comment]
@coderabbitai Outside-diff items fixed in a5b2f6ac:

- 327-331 — `debug_assert!` → `assert!` on accept_data pending guard
- 358-363 — added `assert!(space.len() <= cap)` precondition before the eviction math

Both panic with clear domain-specific messages on protocol / precondition violation instead of silently corrupting state.
